mlir/include/mlir/Dialect/Mesh/IR/MeshBase.td

```diff
@@ -0,0 +1,91 @@
+//===- MeshBase.td - Mesh Dialect --------------------------*- tablegen -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef MLIR_DIALECT_MESH_IR_MESHBASE_TD
+#define MLIR_DIALECT_MESH_IR_MESHBASE_TD
+
+include "mlir/IR/OpBase.td"
+include "mlir/IR/AttrTypeBase.td"
+include "mlir/IR/BuiltinTypeInterfaces.td"
+
+//===----------------------------------------------------------------------===//
+// Mesh Dialect
+//===----------------------------------------------------------------------===//
+
+def Mesh_Dialect : Dialect {
+  let name = "mesh";
+  let cppNamespace = "::mlir::mesh";
+
+  let description = [{
+    The `mesh` dialect contains a set of attributes, operations, interfaces that
+    are useful for representing sharding and communication on device mesh
+    cluster.
+  }];
+
+  let dependentDialects = [
+    "arith::ArithDialect"
+  ];
+
+  let useDefaultAttributePrinterParser = 1;
+  let hasConstantMaterializer = 1;
+}
+
+//===----------------------------------------------------------------------===//
+// Mesh Attribute
+//===----------------------------------------------------------------------===//
+
+def MeshSharding : AttrDef<Mesh_Dialect, "MeshSharding"> {
+  let mnemonic = "shard";
+
+  let parameters = (ins
+    OptionalParameter<"::mlir::SymbolRefAttr">:$cluster,
```

Comment:
Why is this optional?

Reply:
I'm thinking about the default `cluster` symbol ref could be fetched from func op's attributes to simplify the IR text. The downside is that some complexity will be introduced in the implementation of related passes.

I'd like to hear your preference regarding whether the cluster should be optional or not.

Reply:
I'm fine with optional, but we should document it as such!

I wouldn't refer to the function but to an interface, e.g.:

> If the cluster symbol attribute isn't provided, a default cluster will be retrieved from the closest enclosing operation implementing the `DefaultClusterProviderOpInterface`.

Reply:
I just remembered another reason to make the `cluster` optional in mesh sharding attribute is for function reuse, where the cluster info is actually in `func.call` op.

```
func.func @foo() ...

func.func @main() {
    ...
    %1 = func.call @foo(%0) { mesh_cluster = @mesh0 } ...
    %2 = func.call @foo(%1) { mesh_cluster = @mesh1 } ...
    %3 = func.call @foo(%2) { mesh_cluster = @mesh2 } ...
    ...
}
```

This should be useful in pipeline parallel. Here we only need to add one function instead of three. The additional logic can be summarized as follows: Upon encountering an enclosing FunctionInterface operation without any associated cluster information, all corresponding func.call operations will be examined for cluster data.  It may be prudent to introduce some verification logic. If a cluster is identified in one func.call operation, then all func.call operations should contain the cluster information. Meanwhile, all the clusters should have exactly the same rank and `dim_sizes`.

Reply:
As I remember, GSPMD introduces a new dimension in the `axes` of MeshSharding to simplify pipeline parallel. I suggest we could just make `cluster` mandatory in MeshSharding attribute to avoid too much complexity at the early stage.

And I will remove the optional.

Reply:
I'm not totally sure about trying this to an op with FunctionInterface, this seems somehow orthogonal to the notion of a "function".

Also seems like more of a case of "function specialization", where you'll need to clone the function and attach the mesh for each call site with a new mesh...

Reply:
Updated, `optional` is removed.